torch/csrc/jit/script/compiler.cpp

```diff
@@ -1070,44 +1070,65 @@ struct to_ir {
     return emitIfExpr(expr.range(), cond_value, true_expr, false_expr);
   }
 
-  // emit a single expr from the loop comprehension so that we can correctly
-  // type the list we create, then remove the nodes we emitted
-  TypePtr getListCompType(
-      const ListComp& lc,
-      const ListTypePtr& input_list_type) {
+  TypePtr emitInTempEnvironment(const std::function<TypePtr()>& fn) {
     auto b = graph->insertNode(graph->create(prim::Loop))->addBlock();
     pushFrame(b);
     WithInsertPoint guard(b);
-    auto li_elem = graph->insertNode(
-        graph->createUninitialized(input_list_type->getElementType()));
-    emitExprsAssign(
-        List<Expr>::create(lc.range(), {lc.target()}),
-        {std::make_shared<SimpleValue>(li_elem->output())},
-        lc.range(),
-        /*n_binders*/ 1);
-    auto ret_type = emitExpr(lc.elt())->type();
+    auto ret_type = fn();
     popFrame();
     b->owningNode()->destroy();
     return ret_type;
   }
 
+  // emit a single expr from the loop comprehension so that we can correctly
+  // type the list we create, then remove the nodes we emitted
+  TypePtr getListCompType(const ListComp& lc, const TypePtr type_ptr) {
+    return emitInTempEnvironment([&]() {
+      auto li_elem = graph->insertNode(graph->createUninitialized(type_ptr));
+      emitExprsAssign(
+          List<Expr>::create(lc.range(), {lc.target()}),
+          {std::make_shared<SimpleValue>(li_elem->output())},
+          lc.range(),
+          /*n_binders*/ 1);
+      return emitExpr(lc.elt())->type();
+    });
+  }
+
+  // emit a single index of the Iterable to get the children type
+  TypePtr getIterableChildrenType(const ListComp& lc, IterableTree& tree) {
+    return emitInTempEnvironment([&]() {
+      return tree.getitem(lc.range(), method, graph->insertConstant(0))->type();
+    });
+  }
+
   Value* emitListComprehension(const ListComp& lc) {
     const auto tmp_name = createTempName("$list_acc");
-    const auto list_value = emitExpr(lc.iter());
-    if (list_value->type()->kind() != TypeKind::ListType) {
-      // TODO: constraining iterators to be simple lists for now
-      // as it makes easy to get list's element type.
+    const auto sv = emitSugaredExpr(lc.iter(), 1);
+
+    TypePtr list_type;
+
+    auto siv = dynamic_cast<SimpleValue*>(sv.get());
```

Comment:
This code here looks pretty fragile. It the particular way we implement iterators with the implementation of list comprehensions, which are suppose to be syntax sugar. I feel like it would be better to refactor the way we emit lists rather than relying on direct desugaring to ASTs. It isn't necessary to know the type of the list to emit the loop, the list construct and append can simply be added after the loop is emitted. In addition to the coupling, I am not sure it is always safe to emit expressions twice, even if you delete the results.  There may be logic added later that would assume, quite reasonable, that an expression is only emitted once.

Reply:
I'll try to refactor

Reply:
We also emit expressions multiple times because of the tuple-for loop desugaring though

Reply:
But those expressions are part of an unrolling and are used. The weird state is for something potentially side-effectful getting emitted and then not used. That currently doesn't happen (conversions in schema matching are not side-effectful).